File: fourier_transform.py
# Progarm to conduct a fourier transform on a data signal

from math import *
from cmath import *
import numpy as np
import wave
import struct
import matplotlib.pyplot as plt

# Parameters
amp = 16000 # Amplitude
freq = 1000 # Frequency
ns = 32 # number of samples
sr = 32 # sampling rate (for analog to digital conversion)


file = "test.wav"

# Implement structure
sin_wave_uncompressed = [sin(2 * pi * freq * x/sr) for x in range(ns)]
#swu = np.ndarray(shape=(32, 1), dtype = np.float64, buffer = bytes(sin_wave_uncompressed))
# TODO: Play wave uncompressed

# Create structure for a complex exponential function
Fourier = [exp(1j * amp * x/sr) for x in range(ns)]
# Multiply before integration
func = []
for i in range(ns):
    func.append(sin_wave_uncompressed[i] * Fourier[i])
fn = [z.real for z in func]
dn = (min(fn) + max(fn))/len(fn)

fns = np.ndarray(len(func), dtype = complex, buffer = np.array(func))
X_ = [x for x in range(ns)]

wave_compressed = np.trapz(func, X_, dx = dn, axis = -1)

# TODO: Play wave compressed

File: fourier_transform.cc
#include "fourier_transform.h"
#include <vector>
#include <iostream>
#include <cmath>
#include <complex>

using namespace std;
using namespace Data;

vector<complex<double>> wave::transform(vector<double> am, vector<double> fs, int sr, int ns)
{
  complex<double> s;
  vector<complex<double>> result_before_integration;

  for(int j = 0; j < ns; j++)
  {
    s = exp(1i * am.at(j) * j/sr);
    result_before_integration.pushback(s);
  }

  vector<complex<double>> result;
  result = wave::integ(result_before_integration, 0, double(ns), 1.0);
}

vector<complex<double>> wave::integ(vector<complex<double>> v, double x0, double x1, double h)
{
  vector<complex<double>> r;

  for(double j = x0; j < x1; j = j + h)
  {
    complex<double> rr;
    for(double k = x0; k <= j + h; k = k + h)
    {
      complex<double> u = v.at(k);
      double ureal = u.real();
      double uim = u.imag();

      if((k == x0) || (k == j + h))
      {
        rr = rr + (ureal + uim*1i);
      }
      else
      {
        rr = rr + (3 * (ureal + uim*1i));
      }
    }
    rr = 3*rr/8;
    r.pushback(rr);
  }

  return r;
}
